src/Aspire.Dashboard/Components/Pages/Resources.razor.cs

```diff
@@ -116,14 +127,21 @@ await foreach (var (changeType, resource) in subscription.WithCancellation(_watc
                     if (changeType == ResourceViewModelChangeType.Upsert)
                     {
                         _resourceByName[resource.Name] = resource;
+
+                        ImmutableInterlocked.Update(ref _allResourceTypes, set => set.Add(resource.ResourceType));
                     }
                     else if (changeType == ResourceViewModelChangeType.Delete)
                     {
                         var removed = _resourceByName.TryRemove(resource.Name, out _);
                         Debug.Assert(removed, "Cannot remove unknown resource.");
                     }
 
-                    await InvokeAsync(StateHasChanged);
+                    await InvokeAsync(() =>
+                    {
+                        _visibleResourceTypes.Add(resource.ResourceType);
```

Comment:
Does the resource type need to be added inside InvokeAsync?

The change will come from another thread so the hashset should change to a concurrentdictionary (key and value can be the same). Or an immutable collection.

Reply:
_visibleResourceTypes isn't thread safe